### PR TITLE
TypeScript Conversion

### LINKED ISSUE #426: Convert to TypeScript
Convert code to typescript and also set up linting + formatting

### PR DESCRIPTION
**Pull request by ikreymer**
Follows #424. Converts the upcoming 1.0.0 branch based on in-browser recording to TypeScript.
Fixes #426

### CODE REVIEW

**Comment by ikreymer** [comment]
A bit messy, as includes commits from `recorder-work`, but squashing it was a bit tricky

**Review by emma-sg — APPROVED** [review]
All looks good to me!

File: Dockerfile [hunk 1]
```diff
@@ -38,14 +38,20 @@ RUN mkdir -p /tmp/ads && cd /tmp/ads && \
 
 RUN yarn install --network-timeout 1000000
 
-ADD *.js /app/
-ADD util/*.js /app/util/
+#ADD *.js /app/
+#ADD util/*.js /app/util/
```
Comment (tw4l) [inline Dockerfile:42]:
```suggestion
```

File: src/crawler.ts [hunk 2]
```diff
@@ -200,11 +308,16 @@ export class Crawler {
...
+      logger.debug(
+        `Screencast server started on: ${this.params.screencastPort}`,
+        {},
+        "screencast"
+      );
     }
+    // } else if (this.params.redisStoreUrl && this.params.screencastRedis) {
+    //   transport = new RedisPubSubTransport(this.params.redisStoreUrl, this.crawlId);
+    //   logger.debug("Screencast enabled via redis pubsub", {}, "screencast");
+    // }
```
Comment (tw4l) [inline src/crawler.ts:320]: Curious why this is commented out - are we removing this option or is there a follow-up? Option still seems available in argParser

File: src/util/argParser.ts [hunk 3]
```diff
@@ -423,7 +423,7 @@ class ArgParser {
 
       "customBehaviors": {
         describe: "injects a custom behavior file or set of behavior files in a directory",
-        type: ["string"]
+        //type: ["string"]
```
Comment (tw4l) [inline src/util/argParser.ts:426]: Curious why this is commented out, is it causing a typing error? Seeing same for warcInfo above

File: src/util/state.ts [hunk 4]
```diff
@@ -1,60 +1,146 @@
...
+  skey: string;
+  dkey: string;
+  fkey: string;
+  ekey: string;
+  
+  constructor(redis: Redis, key: string, maxPageTime: number, uid: string) {
     this.redis = redis;
 
-    this.maxRetryPending = 1;
+    //this.maxRetryPending = 1;
```
Comment (tw4l) [inline src/util/state.ts:141]:
```suggestion
```

File: src/util/state.ts [hunk 5]
```diff
@@ -1,60 +1,146 @@
...
 
-    this._lastSize = 0;
+    //this._lastSize = 0;
```
Comment (tw4l) [inline src/util/state.ts:143]:
```suggestion
```

**Review by tw4l — APPROVED** [review]
Looks good! Plan is to remove the remaining `any`s incrementally as we work on the codebase, yeah?

File: src/crawler.ts (hunk 2 above)
Comment (ikreymer) [reply src/crawler.ts:320]: Yeah, removing this - it hasn't been used in a long time, and proved to be inefficient compared to direct WS transport, should remove in a follow-up

File: src/util/argParser.ts (hunk 3 above)
Comment (ikreymer) [reply src/util/argParser.ts:426]: ah, it should be `type: "array"`

Comment (ikreymer) [inline src/util/argParser.ts:426]:
```suggestion
        type: "array"
```

**Comment by emma-sg** [comment]
> Looks good! Plan is to remove the remaining `any`s incrementally as we work on the codebase, yeah?

Yep, exactly!

**Pull request merged**